src/net_processing.cpp

```diff
@@ -3334,6 +3388,58 @@ bool PeerLogicValidation::SendMessages(CNode* pto, std::atomic<bool>& interruptM
             }
         }
 
+        // TODO: This logic is not peer-specific, so it doesn't really make
+        // sense in SendMessages; move this somewhere more fitting.
+        // Our goal here is to avoid being connected to a set of outbound peers
+        // which are all refusing to relay valid blocks to us.
+        // Strategy: if too much time has passed since we last updated our tip,
+        // mark the outbound peer with the least recent block announcement as
+        // evictable by our net.cpp code (which will only evict if we are at
+        // our outbound peer limit).
+        // If multiple peers are tied for least recently announcing a
+        // block to us, prefer to evict the newest peer -- this prevents us
+        // from cycling through all our peers in the event that blocks are just
+        // slow to be found on our chain.
+        if (time_in_seconds > g_tip_stale_check_time) {
+            if (TipMayBeStale(consensusParams)) {
+                // Mark our "worst" outbound peer for potential eviction, in
+                // the hopes of finding an outbound peer who has a new block.
+                NodeId worst_peer = -1;
+                int64_t oldest_block_announcement = GetTime();
+                int64_t worst_peer_connect_time = 0;
+                for (auto it = mapNodeState.begin(); it != mapNodeState.end(); ++it) {
+                    int64_t connect_time = 0;
+                    if (connman->ForNode(it->first, [&](CNode *pnode){
```

Comment:
I wonder if we might turn this logic around: rather than marking for disconnection, mark for protection from disconnection. For example, each time we get a new header that extends our tip, we could do:
```
connman->ResetOutboundEvictionTimer(id, current_time + new_timeout);
```
Then when we need a connection slot, if any timer has expred, whichever has been expired the longest gets dropped.
That saves us from having to care about what type of node (inbound/outbound/oneshot/manual/etc.) this is, and just lets outbound eviction do its thing if the criteria are met.